### DISCUSSION TITLE
CPU Core Utilization in HPCGAP/WSL

### DISCUSSION THREAD

**Discussion by jaslamMiktex** [gap-system/gap#5496] [Ideas]
I have a working program on HPCGAP (using Ryzen 7- 8-core CPU), using WSL/HPCGAP build
But I do not see more than 2 cores being used in the WSL.
Here is the prototype template of my real application (real app not shown):
```
############################
createTasks := function(n)
local i, t, tr, sum, sumx;
   t := [];
   arrayDB[1] := [1 .. n];
   sumx := [];
   for i in [1 ..n ] do
      t[i] := RunTask(taskFunction,  i);
   od;
  sum := [];

   for i in [1 .. n] do
     WaitTask(t[i]);
     Add(sum, arrayDB[i]);
   od;
 return sum[Length(sum)];
 end;
#-------------------
taskFunction := function(r)
 local k, arx;
 arx := arrayDB[r];
 arrayDB[r+1] := List(arx, x->x+1);
# return arrayDB[r+1];
 return;
 end;
##############################
```
Any ideas will be highly appreciated.

Thank you

**Comment by fingolfin** [comment]
What does `GAPInfo.KernelInfo.NUM_CPUS` report? It may have not recognized the number of CPU cores correctly.

You could use use `SetMaxTaskWorkers` to get it use more worker threads, e.g. `SetMaxTaskWorkers(12)` to use up to 12 worker threads.

**Reply by jaslamMiktex** [reply]
Thank you Max. Will try that.

**Reply by jaslamMiktex** [reply]
Latest trial:
Finally, I was able to run this HPC-GAP application with 100% CPU utilization on all the cores, giving a 3X improvement
in the running time.
SetMaxTaskWorkers (4) or SetMaxTaskWorkers(16) does not make more 10% difference in the performance (running time).
At the startup, HPC GAP declares that only "4 worker threads are being used." 
PS: With all the (hopefully) correct settings like ShareObj(), etc, I continued into the "read-access" violation problem.
The only resolution I found was using DISABLE_GUARDS(2);
<---------------------------------------------------------  Thank you

**Reply by ChrisJefferson** [reply]
I just want to point out that disabling guards turns off various checks. While your calculation may produce the correct answer, there is also a chance you might get very wrong answers -- I would not trust any generated result unless you have an independent way to verify it (having some way of checking your answer is of course always a good idea when possible, but particularly in this case).

**Reply by jaslamMiktex** [reply]
Yes, I do have a sequential version (using GAP 4.12.2) as well to verify. 
There is no shared write area, so I do not see why would that be a problem. Those read-only data structures remain read-only-- never modified.  Almost like a software ROM module.
